File: hmd_agro/hmd_agro/page/import_traites/import_traites.py
import frappe
import openpyxl
from frappe.utils import cint, getdate, today
from frappe.utils.background_jobs import enqueue, is_job_enqueued
from io import BytesIO

from hmd_agro.hmd_agro.utils.config import get_config


@frappe.whitelist()
def preview_import(file_url):
    """Parse Excel and return preview: matched animals, date range, row counts"""
    dates, animals_data = parse_excel(file_url)

    if not dates:
        frappe.throw("Aucune date trouvee dans le fichier (ligne 1, colonnes B+).")
    if not animals_data:
        frappe.throw("Aucun animal trouve dans le fichier (colonne A).")

    # Build animal mapping: nom_metier -> animal name
    animal_map, duplicate_noms = build_animal_mapping()

    # Match Excel animals to DB
    matched = []
    unmatched = []
    duplicates = []
    seen_duplicates = set()

    for row in animals_data:
        nom = str(row["nom_metier"]).zfill(4)
        if nom in duplicate_noms:
            if nom not in seen_duplicates:
                duplicates.append(nom)
                seen_duplicates.add(nom)
        elif nom in animal_map:
            matched.append({"nom_metier": nom, "animal": animal_map[nom]})
        else:
            unmatched.append(nom)

    # Build candidate list for each duplicate nom_metier
    duplicate_candidates = {}
    for nom in duplicates:
        candidates = frappe.db.get_all("Animal",
            filters={"nom_metier": nom},
            fields=["name", "identification_tn", "identification_fr",
                    "id_lot", "statut", "categorie"]
        )
        duplicate_candidates[nom] = candidates

    # Build lactation info for matched animals
    lactation_info = {}
    for m in matched:
        lacs = get_animal_lactations(m["animal"])
        lactation_info[m["nom_metier"]] = {
            "animal": m["animal"],
            "lactations": len(lacs),
            "details": [
                {
                    "name": l.name,
                    "numero": l.numero_lactation,
                    "debut": str(l.date_debut),
                    "fin": str(l.date_tarissement) if l.date_tarissement else "en cours",
                    "statut": l.statut
                }
                for l in lacs
            ]
        }

    return {
        "date_range": {
            "start": str(dates[0]),
            "end": str(dates[-1]),
            "days": len(dates)
        },
        "total_animals_excel": len(animals_data),
        "matched": len(matched),
        "unmatched": unmatched,
        "duplicates": duplicates,
        "duplicate_candidates": duplicate_candidates,
        "estimated_traites": len(matched) * len(dates) * 2,
        "lactation_info": lactation_info
    }


def _result_key(file_url):
    return f"import_traites_result::{file_url}"


@frappe.whitelist()
def run_import(file_url, keep_original=1, resolutions=None):
    """Enqueue the import as a background job (a full-year file takes ~90s — too long for a
    synchronous web request, which the proxy times out). The job stashes its result in cache;
    the page polls get_import_result() for it. Polling, not realtime — background-job realtime
    delivery proved unreliable, so the result never surfaced."""
    import json
    keep_original = int(keep_original)
    if isinstance(resolutions, str):
        resolutions = json.loads(resolutions) if resolutions else {}
    resolutions = resolutions or {}

    frappe.cache().delete_value(_result_key(file_url))   # clear any prior result
    # Always enqueue with an RQ-assigned unique id — a fixed job_id could linger in the
    # finished/started registry and make a re-import of the same file silently not run.
    enqueue(
        _process_import, queue="long", timeout=6000,
        file_url=file_url, keep_original=keep_original,
        resolutions=resolutions, user=frappe.session.user,
        now=frappe.conf.developer_mode,
    )
    return {"status": "started", "file_url": file_url}


@frappe.whitelist()
def get_import_result(file_url):
    """Poll endpoint: returns the stored result once the background import finishes, else None."""
    return frappe.cache().get_value(_result_key(file_url))


def _process_import(file_url, keep_original, user, resolutions=None):
    """Background job: parse Excel, create Traite records, stash the result in cache."""
    frappe.set_user(user)
    resolutions = resolutions or {}

    try:
        dates, animals_data = parse_excel(file_url)

        if not dates or not animals_data:
            frappe.cache().set_value(_result_key(file_url),
                {"success": False, "error": "Fichier invalide."}, expires_in_sec=3600)
            return

        animal_map, duplicate_noms = build_animal_mapping()

        # Apply user resolutions for ambiguous nom_metier
        for nom, animal_name in resolutions.items():
            if animal_name:
                animal_map[nom] = animal_name
                duplicate_noms.discard(nom)

        summary = {
            "created": 0,
            "overwritten": 0,
            "skipped_no_animal": 0,
            "skipped_no_lactation": 0,
            "skipped_after_sortie": 0,
            "skipped_duplicate": 0,
            "errors": [],
            "lactations_updated": 0
        }

        affected_lactations = set()

        # Calculate total for progress
        total_cells = len(animals_data) * len(dates)
        processed = 0

        for row in animals_data:
            nom = str(row["nom_metier"]).zfill(4)

            # Skip unmatched or duplicate nom_metier
            if nom in duplicate_noms or nom not in animal_map:
                summary["skipped_no_animal"] += len(dates)
                processed += len(dates)
                frappe.publish_realtime("import_traites_progress",
                    {"current": processed, "total": total_cells},
                    user=user)
                continue

            animal_name = animal_map[nom]
            lactations = get_animal_lactations(animal_name)
            animal_lot = frappe.db.get_value("Animal", animal_name, "id_lot")
            # Mirror Traite.validate_animal_present_on_date — refuse to import
            # traites whose date is after the animal's exit. Without this,
            # Excel files containing stale rows for sold/dead cows silently
            # create post-sortie traites (140 such rows found in the audit
            # before this guard was added).
            animal_sortie = frappe.db.get_value("Animal", animal_name, "date_sortie")

            for i, date in enumerate(dates):
                value = row["values"][i] if i < len(row["values"]) else 0
                try:
                    value = float(value) if value else 0
                except (ValueError, TypeError):
                    summary["errors"].append({
                        "animal": nom,
                        "date": str(date),
                        "reason": f"Valeur non numerique: {value}"
                    })
                    processed += 1
                    continue

                # Skip rows whose date is after the animal's exit (same rule as
                # Traite.validate_animal_present_on_date).
                if animal_sortie and getdate(date) > getdate(animal_sortie):
                    summary["skipped_after_sortie"] += 1
                    if value > 0:
                        summary["errors"].append({
                            "animal": nom,
                            "date": str(date),
                            "reason": f"Animal sorti le {animal_sortie} — "
                                      f"date postérieure ignorée"
                        })
                    processed += 1
                    continue

                # Find which lactation covers this date
                lactation = find_lactation_for_date(lactations, date)

                if not lactation:
                    summary["skipped_no_lactation"] += 1
                    if value > 0:
                        summary["errors"].append({
                            "animal": nom,
                            "date": str(date),
                            "reason": "Pas de lactation pour cette date"
                        })
                    processed += 1
                    continue

                # 50/50 split
                half = round(value / 2, 1)

                max_litres = get_config("traite_max_litres", default=60)
                for session in ["MATIN", "SOIR"]:
                    # Validate quantity (0..max_litres per session — same limit as Traite.validate_quantite)
                    qty = half
                    if qty > max_litres:
                        summary["errors"].append({
                            "animal": nom,
                            "date": str(date),
                            "reason": f"Quantite {session} depasse {max_litres}L ({qty})"
                        })
                        continue

                    # Check duplicate
                    existing = frappe.db.exists("Traite", {
                        "animal": animal_name,
                        "date_traite": str(date),
                        "session": session
                    })
                    if existing:
                        if keep_original:
                            summary["skipped_duplicate"] += 1
                            continue
                        else:
                            # Override: update existing traite. Reset brut to the
                            # imported value too — the import is the new measurement.
                            frappe.db.set_value("Traite", existing,
                                {"quantite_litres": qty,
                                 "quantite_litres_brut": qty,
                                 "lactation": lactation.name})
                            summary["overwritten"] += 1
                            affected_lactations.add(lactation.name)
                            continue

                    doc = frappe.get_doc({
                        "doctype": "Traite",
                        "animal": animal_name,
                        "date_traite": str(date),
                        "session": session,
                        "quantite_litres": qty,
                        "quantite_litres_brut": qty,
                        "lactation": lactation.name,
                        "id_lot": animal_lot
                    })
                    doc.flags.ignore_validate = True
                    doc.flags.ignore_links = True
                    doc.flags.skip_lactation_update = True  # final recalc done once at end of import
                    doc.insert(ignore_permissions=True)
                    summary["created"] += 1
                    affected_lactations.add(lactation.name)

                processed += 1

                # Publish progress every 50 cells
                if processed % 50 == 0:
                    frappe.publish_realtime("import_traites_progress",
                        {"current": processed, "total": total_cells},
                        user=user)

            # Commit per animal
            frappe.db.commit()

        # Recalculate production for all affected lactations
        for lac_name in affected_lactations:
            recalculate_lactation_production(lac_name)
        frappe.db.commit()

        summary["lactations_updated"] = len(affected_lactations)
        frappe.cache().set_value(_result_key(file_url),
            {"success": True, "summary": summary}, expires_in_sec=3600)

    except Exception:
        frappe.db.rollback()
        frappe.log_error("Import traites failed")
        frappe.cache().set_value(_result_key(file_url),
            {"success": False, "error": str(frappe.get_traceback())}, expires_in_sec=3600)


def find_lactation_for_date(lactations, date):
    """Find the lactation that covers the given date"""
    for lac in lactations:
        if not lac.date_debut:
            continue
        debut = getdate(lac.date_debut)
        if debut > date:
            continue
        # EN_COURS has no end date — covers up to today
        if not lac.date_tarissement:
            if date <= getdate(today()):
                return lac
        else:
            if date <= getdate(lac.date_tarissement):
                return lac
    return None


def recalculate_lactation_production(lactation_name):
    """Recalculate production totals for a lactation (same logic as Traite.update_lactation_production)"""
    date_debut = frappe.db.get_value("Lactation", lactation_name, "date_debut")

    total = frappe.db.sql("""
        SELECT SUM(quantite_litres)
        FROM `tabTraite`
        WHERE lactation = %s
    """, lactation_name)[0][0] or 0

    pic_window = cint(get_config("pic_production_jours", default=150))
    pic = frappe.db.sql(f"""
        SELECT MAX(daily_total) FROM (
            SELECT SUM(quantite_litres) as daily_total
            FROM `tabTraite`
            WHERE lactation = %s
              AND DATEDIFF(date_traite, %s) <= {pic_window}
            GROUP BY date_traite
        ) as daily
    """, (lactation_name, date_debut))[0][0] or 0

    updates = {
        "production_totale": total,
        "pic_production": pic
    }

    if date_debut:
        total_305 = frappe.db.sql("""
            SELECT SUM(quantite_litres)
            FROM `tabTraite`
            WHERE lactation = %s
              AND DATEDIFF(date_traite, %s) <= 305
        """, (lactation_name, date_debut))[0][0] or 0
        updates["lactation_305j"] = round(total_305, 2)

        init_window = cint(get_config("production_initiale_jours", default=60))
        prod_init = frappe.db.sql(f"""
            SELECT SUM(quantite_litres)
            FROM `tabTraite`
            WHERE lactation = %s
              AND DATEDIFF(date_traite, %s) <= {init_window}
        """, (lactation_name, date_debut))[0][0] or 0
        updates["production_initiale"] = round(prod_init, 2)

        from frappe.utils import date_diff
        jours = date_diff(today(), date_debut)
        if jours > 0:
            updates["moyenne_production"] = round(total / jours, 2)

    frappe.db.set_value("Lactation", lactation_name, updates)


def parse_excel(file_url):
    """Parse the uploaded Excel file, return (dates, animals_data)"""
    file_doc = frappe.get_doc("File", {"file_url": file_url})
    content = file_doc.get_content()
    wb = openpyxl.load_workbook(BytesIO(content), data_only=True)
    ws = wb.active

    # Row 1: date headers starting at col B (index 2)
    dates = []
    for col in range(2, ws.max_column + 1):
        val = ws.cell(1, col).value
        if val and hasattr(val, 'date'):
            dates.append(val.date())
        elif val and isinstance(val, str):
            try:
                dates.append(getdate(val))
            except Exception:
                break
        else:
            break

    # Rows 2+: col A = nom_metier, cols B+ = daily values
    animals_data = []
    for row in range(2, ws.max_row + 1):
        nom_metier = ws.cell(row, 1).value
        if nom_metier is None:
            continue

        values = []
        for col_idx in range(2, 2 + len(dates)):
            val = ws.cell(row, col_idx).value
            values.append(val if val is not None else 0)

        animals_data.append({
            "nom_metier": str(nom_metier).zfill(4) if isinstance(nom_metier, int) else str(nom_metier),
            "values": values
        })

    return dates, animals_data


def build_animal_mapping():
    """Map a milk-file animal number to an Animal by EITHER identification_fr (N°Fr)
    or nom_metier (TN suffix) — both zero-padded to 4 to match the file parsing.
    So a file can be keyed by French OR Tunisian numbers (even mixed in one column).

    A number that resolves to >1 distinct animal (e.g. one cow's FR equals another
    cow's TN suffix) is AMBIGUOUS -> returned in `duplicates` so the import's
    resolution dialog asks the user which cow it is (never silently wrong).
    Returns (map, duplicates set)."""
    from collections import defaultdict
    animals = frappe.db.get_all("Animal",
        fields=["name", "nom_metier", "identification_fr"])

    key_to_animals = defaultdict(set)
    for a in animals:
        for val in (a.nom_metier, a.identification_fr):
            if val:
                key_to_animals[str(val).zfill(4)].add(a.name)

    duplicates = {k for k, names in key_to_animals.items() if len(names) > 1}
    mapping = {k: next(iter(names)) for k, names in key_to_animals.items()
               if len(names) == 1}

    return mapping, duplicates


def get_animal_lactations(animal_name):
    """Get all lactations for an animal, sorted by date_debut"""
    return frappe.db.get_all("Lactation",
        filters={"animal": animal_name},
        fields=["name", "numero_lactation", "date_debut", "date_tarissement", "statut"],
        order_by="date_debut asc"
    )

File: hmd_agro/hmd_agro/page/import_traites/import_traites.js
frappe.pages["import-traites"].on_page_load = function (wrapper) {
    let page = frappe.ui.make_app_page({
        parent: wrapper,
        title: "Import Traites",
        single_column: true,
    });

    page.main.html(`
        <div class="import-traites-container" style="max-width:900px; margin:0 auto; padding:20px;">
            <div class="upload-section" style="border:1px solid var(--border-color); border-radius:8px; padding:20px; background:var(--card-bg);">
                <h5>1. Charger le fichier Excel</h5>
                <p style="color:var(--text-muted); font-size:13px;">
                    Format attendu: Ligne 1 = dates (colonnes B+), Colonne A = nom metier, cellules = production journaliere (litres)
                </p>
                <div class="file-upload-area"></div>
                <div style="margin-top:15px;">
                    <button class="btn btn-primary btn-preview">Analyser le fichier</button>
                </div>
            </div>

            <div class="preview-section" style="display:none; margin-top:20px;">
                <div class="preview-content"></div>
                <div style="margin-top:15px; display:flex; align-items:center; gap:15px;">
                    <button class="btn btn-primary btn-sm btn-import">Lancer l'import</button>
                    <label style="font-size:13px; cursor:pointer; margin:0;">
                        <input type="checkbox" class="chk-keep-original" checked style="margin-right:5px;">
                        Garder les traites existantes
                    </label>
                </div>
            </div>

            <div class="results-section" style="display:none; margin-top:20px;">
                <div class="results-content"></div>
            </div>
        </div>
    `);

    // File upload control
    let file_control = frappe.ui.form.make_control({
        parent: page.main.find(".file-upload-area"),
        df: {
            fieldtype: "Attach",
            fieldname: "import_file",
            label: "Fichier Excel",
            options: { restrictions: { allowed_file_types: [".xlsx", ".xls"] } }
        },
        render_input: true,
    });

    page.file_control = file_control;

    // Preview button
    page.main.find(".btn-preview").on("click", function () {
        let file_url = file_control.get_value();
        if (!file_url) {
            frappe.msgprint("Veuillez charger un fichier Excel.");
            return;
        }

        frappe.call({
            method: "hmd_agro.hmd_agro.page.import_traites.import_traites.preview_import",
            args: { file_url: file_url },
            freeze: true,
            freeze_message: "Analyse du fichier...",
            callback: function (r) {
                if (r.message) {
                    render_preview(page, r.message);
                }
            },
        });
    });

    // Import button
    page.main.find(".btn-import").on("click", function () {
        let file_url = file_control.get_value();
        if (!file_url) {
            frappe.msgprint("Veuillez charger un fichier Excel.");
            return;
        }

        // Collect duplicate-resolution selections
        let resolutions = {};
        page.main.find(".duplicate-resolution-select").each(function () {
            let nom = $(this).data("nom");
            let val = $(this).val();
            if (val) {
                resolutions[nom] = val;
            }
        });

        let keep_original = page.main.find(".chk-keep-original").is(":checked");
        let confirm_msg = "Lancer l'import ? Cette operation va creer les traites pour tous les animaux trouves.";
        if (!keep_original) {
            confirm_msg = "<strong>Attention :</strong> Les traites existantes seront ecrasees par les valeurs du fichier Excel.<br><br>Voulez-vous continuer ?";
        }

        frappe.confirm(
            confirm_msg,
            function () {
                // Show progress section
                let $results = page.main.find(".results-section");
                $results.show();
                $results.find(".results-content").html(`
                    <div style="border:1px solid var(--border-color); border-radius:8px; padding:20px; background:var(--card-bg);">
                        <h5>Import en cours...</h5>
                        <div class="progress" style="margin-top:15px;">
                            <div class="progress-bar progress-bar-striped progress-bar-animated"
                                 role="progressbar" style="width:0%">0%</div>
                        </div>
                        <div class="import-status" style="margin-top:10px; font-size:13px; color:var(--text-muted);"></div>
                    </div>
                `);

                // Disable import button
                page.main.find(".btn-import").prop("disabled", true);

                // Listen for progress
                frappe.realtime.on("import_traites_progress", function (data) {
                    let pct = Math.round((data.current / data.total) * 100);
                    $results.find(".progress-bar").css("width", pct + "%").text(pct + "%");
                    $results.find(".import-status").text(
                        data.current + " / " + data.total + " cellules traitees"
                    );
                });

                // Listen for completion
                frappe.realtime.on("import_traites_complete", function (data) {
                    frappe.realtime.off("import_traites_progress");
                    frappe.realtime.off("import_traites_complete");
                    page.main.find(".btn-import").prop("disabled", false);

                    if (data.success) {
                        render_results(page, data.summary);
                    } else {
                        $results.find(".results-content").html(`
                            <div style="border:1px solid var(--border-color); border-radius:8px; padding:20px; background:var(--alert-bg-danger);">
                                <h5>Erreur</h5>
                                <p>${data.error || "Une erreur est survenue."}</p>
                            </div>
                        `);
                    }
                });

                // Start import
                frappe.call({
                    method: "hmd_agro.hmd_agro.page.import_traites.import_traites.run_import",
                    args: {
                        file_url: file_url,
                        keep_original: keep_original ? 1 : 0,
                        resolutions: JSON.stringify(resolutions)
                    },
                });
            }
        );
    });
};

function render_preview(page, data) {
    let $section = page.main.find(".preview-section");
    $section.show();

    let html = `
    <div style="border:1px solid var(--border-color); border-radius:8px; padding:20px; background:var(--card-bg);">
        <h5>2. Analyse du fichier</h5>

        <div style="display:flex; gap:20px; flex-wrap:wrap; margin-bottom:15px;">
            <div style="flex:1; min-width:150px; padding:10px; border:1px solid var(--border-color); border-radius:6px;">
                <div style="font-size:11px; color:var(--text-muted); text-transform:uppercase;">Periode</div>
                <div style="font-size:16px; font-weight:600;">${data.date_range.start} → ${data.date_range.end}</div>
                <div style="font-size:12px; color:var(--text-muted);">${data.date_range.days} jours</div>
            </div>
            <div style="flex:1; min-width:150px; padding:10px; border:1px solid var(--border-color); border-radius:6px;">
                <div style="font-size:11px; color:var(--text-muted); text-transform:uppercase;">Animaux dans le fichier</div>
                <div style="font-size:16px; font-weight:600;">${data.total_animals_excel}</div>
                <div style="font-size:12px; color:var(--text-muted);">${data.matched} trouves, ${data.unmatched.length} non trouves</div>
            </div>
            <div style="flex:1; min-width:150px; padding:10px; border:1px solid var(--border-color); border-radius:6px;">
                <div style="font-size:11px; color:var(--text-muted); text-transform:uppercase;">Traites estimees</div>
                <div style="font-size:16px; font-weight:600;">${data.estimated_traites}</div>
                <div style="font-size:12px; color:var(--text-muted);">${data.matched} animaux x ${data.date_range.days} jours x 2 sessions</div>
            </div>
        </div>`;

    // Unmatched animals warning
    if (data.unmatched.length > 0) {
        html += `
        <div style="background:var(--alert-bg-warning); border:1px solid var(--alert-text-warning); border-radius:6px; padding:10px; margin-bottom:15px;">
            <strong>Animaux non trouves (${data.unmatched.length}):</strong>
            ${data.unmatched.join(", ")}
        </div>`;
    }

    if (data.duplicates.length > 0) {
        html += `
        <div style="background:var(--alert-bg-warning); border:1px solid var(--alert-text-warning); border-radius:6px; padding:12px; margin-bottom:15px;">
            <strong>N° Travail ambigu (${data.duplicates.length}) — plusieurs animaux partagent ces 4 derniers chiffres :</strong>
            <div style="font-size:12px; color:var(--text-muted); margin-top:4px;">
                Choisissez l'animal a utiliser pour chaque ligne. Les noms non resolus seront ignores.
            </div>
            <table style="width:100%; font-size:12px; border-collapse:collapse; margin-top:10px;" class="duplicate-resolution-table">
                <tr style="border-bottom:1px solid var(--border-color);">
                    <th style="padding:6px 4px; text-align:left;">N° Travail</th>
                    <th style="padding:6px 4px; text-align:left;">Animal a utiliser</th>
                </tr>`;

        data.duplicates.forEach(function (nom) {
            let candidates = data.duplicate_candidates[nom] || [];
            let options = `<option value="">-- Ignorer --</option>`;
            candidates.forEach(function (c) {
                let label_parts = [c.name];
                if (c.identification_fr) label_parts.push("FR:" + c.identification_fr);
                if (c.id_lot) label_parts.push("Lot:" + c.id_lot);
                if (c.categorie) label_parts.push(c.categorie);
                if (c.statut && c.statut !== "ACTIF") label_parts.push("[" + c.statut + "]");
                options += `<option value="${c.name}">${label_parts.join(" — ")}</option>`;
            });

            html += `
            <tr style="border-bottom:1px solid var(--light-border-color);">
                <td style="padding:6px 4px; font-weight:600;">${nom}</td>
                <td style="padding:6px 4px;">
                    <select class="form-control input-xs duplicate-resolution-select"
                            data-nom="${nom}"
                            style="font-size:12px; height:28px;">
                        ${options}
                    </select>
                </td>
            </tr>`;
        });

        html += `</table></div>`;
    }

    // Matched animals with lactation info
    if (data.matched > 0) {
        html += `
        <h6>Animaux trouves avec leurs lactations</h6>
        <table style="width:100%; font-size:12px; border-collapse:collapse;">
            <tr style="border-bottom:1px solid var(--border-color);">
                <th style="padding:6px 4px; text-align:left;">N° Travail</th>
                <th style="padding:6px 4px; text-align:left;">ID Animal</th>
                <th style="padding:6px 4px; text-align:center;">Lactations</th>
                <th style="padding:6px 4px; text-align:left;">Periodes</th>
            </tr>`;

        for (let [nom, info] of Object.entries(data.lactation_info)) {
            let lac_details = info.details.map(
                (l) => `L${l.numero} (${l.debut} → ${l.fin})`
            ).join(", ");

            html += `
            <tr style="border-bottom:1px solid var(--light-border-color);">
                <td style="padding:6px 4px;">${nom}</td>
                <td style="padding:6px 4px;">${info.animal}</td>
                <td style="padding:6px 4px; text-align:center;">${info.lactations}</td>
                <td style="padding:6px 4px; font-size:11px;">${lac_details || '<span style="color:var(--text-muted);">Aucune</span>'}</td>
            </tr>`;
        }
        html += `</table>`;
    }

    html += `</div>`;

    $section.find(".preview-content").html(html);

    // Store preview data for import step
    page.__preview_data = data;
}


function render_results(page, data) {
    let $section = page.main.find(".results-section");
    $section.show();

    let html = `
    <div style="border:1px solid var(--border-color); border-radius:8px; padding:20px; background:var(--card-bg);">
        <h5>3. Resultats de l'import</h5>

        <div style="display:flex; gap:20px; flex-wrap:wrap; margin-bottom:15px;">
            <div style="flex:1; min-width:120px; padding:10px; border:1px solid var(--border-color); border-radius:6px; background:var(--alert-bg-success);">
                <div style="font-size:11px; color:var(--text-muted); text-transform:uppercase;">Creees</div>
                <div style="font-size:22px; font-weight:600; color:var(--alert-text-success);">${data.created}</div>
            </div>
            <div style="flex:1; min-width:120px; padding:10px; border:1px solid var(--border-color); border-radius:6px;">
                <div style="font-size:11px; color:var(--text-muted); text-transform:uppercase;">Ignorees (pas de lactation)</div>
                <div style="font-size:22px; font-weight:600;">${data.skipped_no_lactation}</div>
            </div>
            <div style="flex:1; min-width:120px; padding:10px; border:1px solid var(--border-color); border-radius:6px;">
                <div style="font-size:11px; color:var(--text-muted); text-transform:uppercase;">Ignorees (animal inconnu)</div>
                <div style="font-size:22px; font-weight:600;">${data.skipped_no_animal}</div>
            </div>
            <div style="flex:1; min-width:120px; padding:10px; border:1px solid var(--border-color); border-radius:6px;">
                <div style="font-size:11px; color:var(--text-muted); text-transform:uppercase;">Ecrasees</div>
                <div style="font-size:22px; font-weight:600;">${data.overwritten || 0}</div>
            </div>
            <div style="flex:1; min-width:120px; padding:10px; border:1px solid var(--border-color); border-radius:6px;">
                <div style="font-size:11px; color:var(--text-muted); text-transform:uppercase;">Traites deja existantes</div>
                <div style="font-size:22px; font-weight:600;">${data.skipped_duplicate}</div>
            </div>
            <div style="flex:1; min-width:120px; padding:10px; border:1px solid var(--border-color); border-radius:6px;">
                <div style="font-size:11px; color:var(--text-muted); text-transform:uppercase;">Lactations recalculees</div>
                <div style="font-size:22px; font-weight:600;">${data.lactations_updated}</div>
            </div>
        </div>`;

    if (data.errors && data.errors.length > 0) {
        html += `
        <h6>Erreurs (${data.errors.length})</h6>
        <div style="max-height:300px; overflow-y:auto;">
        <table style="width:100%; font-size:12px; border-collapse:collapse;">
            <tr style="border-bottom:1px solid var(--border-color);">
                <th style="padding:4px;">Animal</th>
                <th style="padding:4px;">Date</th>
                <th style="padding:4px;">Raison</th>
            </tr>`;

        data.errors.forEach(function (e) {
            html += `
            <tr style="border-bottom:1px solid var(--light-border-color);">
                <td style="padding:4px;">${e.animal}</td>
                <td style="padding:4px;">${e.date}</td>
                <td style="padding:4px;">${e.reason}</td>
            </tr>`;
        });

        html += `</table></div>`;
    }

    html += `</div>`;
    $section.find(".results-content").html(html);
}
